# Aula 29

In [1]:
import pandas as pd
import numpy as np

In [2]:
dados_p = pd.read_csv('./p_data.csv',sep = ';')
dados_p

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

In [3]:
dados_w = pd.read_csv('./w_data.csv',sep = ';')
dados_w

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

# Quantos ID têm em comum? Nomear quais não têm em comum.


In [219]:
new_df = dados_w.merge(dados_p, on = 'ID', how = 'inner')
new_df

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48837            Private  245211.0           0.0           0.0   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   

       Hours per week          ID   Target   Age      Sex  \
0                40.0       49084    <=50K  39.0     Male   
1                13.0       68553    <=50K  50.0     Male   
2                40.0      143654    <=50K  38.0     Male   
3                40.0       41510    <=50K  53.0     Male   
4                40.0      195025    <=50K  28.0   Female   
...               ...         ...      ...   ...      ...   
48837            40.0   597186815   <=50K.  33.0     Male   
48838            36.0   735617153   <=50K.  39.0   Female   
48839            40.0  1131807446   <=50K.  64.0     Male   
48840            50.0   781422087   <=50K.  38.0     Male   
48841            40.0   161771296   <=50K.  44.0     Male   

                      Race         Country   Education  Education-Num  \
0                    White   United-States   Bachelors           13.0   
1                    White   United-States   Bachelors           13.0   
2                    White   United-States     HS-grad            9.0   
3                    Black   United-States        11th            7.0   
4                    Black            Cuba   Bachelors           13.0   
...                    ...             ...         ...            ...   
48837                White   United-States   Bachelors           13.0   
48838                White   United-States   Bachelors           13.0   
48839                Black   United-States     HS-grad            9.0   
48840                White   United-States   Bachelors           13.0   
48841   Asian-Pac-Islander   United-States   Bachelors           13.0   

            Marital Status     Relationship  
0            Never-married    Not-in-family  
1       Married-civ-spouse          Husband  
2                 Divorced    Not-in-family  
3       Married-civ-spouse          Husband  
4       Married-civ-spouse             Wife  
...                    ...              ...  
48837        Never-married        Own-child  
48838             Divorced    Not-in-family  
48839              Widowed   Other-relative  
48840   Married-civ-spouse          Husband  
48841             Divorced        Own-child  

[48842 rows x 15 columns]

R: 48842

Resposta : Total de Registros 48842

In [220]:
new_df2 = dados_w.merge(dados_p, on = 'ID', how = 'left')
 
new_df2[new_df2['Age'].isna()]

Workclass    fnlwgt  Capital Gain  Capital Loss  Hours per week  \
32561            NaN       NaN           NaN           NaN             NaN   
48842   Self-emp-inc  182148.0           0.0           0.0            60.0   

               ID  Target  Age  Sex Race Country Education  Education-Num  \
32561       37078     NaN  NaN  NaN  NaN     NaN       NaN            NaN   
48842  1344995623   >50K.  NaN  NaN  NaN     NaN       NaN            NaN   

      Marital Status Relationship  
32561            NaN          NaN  
48842            NaN          NaN

Resposta : Sao 2 registros que nao tem num e nem no outro  ID: 37078 e 1344995623

# 2. Corrigir os target a apenas dois valores distintos ‘<=50k’ e ‘>50k’

In [242]:
dados_geral =  pd.merge(dados_w,dados_p, on ='ID')

pd.Series(dados_geral['Target'], name='A').unique() #verificação dos valores distindos da coluna 'Target'


array([' <=50K', ' >50K', nan, ' <=50K.', ' >50K.'], dtype=object)

In [243]:
dados_geral['Target'] = dados_geral['Target'].map({' >50K.':' >50K',
                                             ' <=50K.':' <=50K'},
                                              na_action=None)

 
dados_geral

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48837            Private  245211.0           0.0           0.0   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   

       Hours per week          ID  Target   Age      Sex                 Race  \
0                40.0       49084     NaN  39.0     Male                White   
1                13.0       68553     NaN  50.0     Male                White   
2                40.0      143654     NaN  38.0     Male                White   
3                40.0       41510     NaN  53.0     Male                Black   
4                40.0      195025     NaN  28.0   Female                Black   
...               ...         ...     ...   ...      ...                  ...   
48837            40.0   597186815   <=50K  33.0     Male                White   
48838            36.0   735617153   <=50K  39.0   Female                White   
48839            40.0  1131807446   <=50K  64.0     Male                Black   
48840            50.0   781422087   <=50K  38.0     Male                White   
48841            40.0   161771296   <=50K  44.0     Male   Asian-Pac-Islander   

              Country   Education  Education-Num       Marital Status  \
0       United-States   Bachelors           13.0        Never-married   
1       United-States   Bachelors           13.0   Married-civ-spouse   
2       United-States     HS-grad            9.0             Divorced   
3       United-States        11th            7.0   Married-civ-spouse   
4                Cuba   Bachelors           13.0   Married-civ-spouse   
...               ...         ...            ...                  ...   
48837   United-States   Bachelors           13.0        Never-married   
48838   United-States   Bachelors           13.0             Divorced   
48839   United-States     HS-grad            9.0              Widowed   
48840   United-States   Bachelors           13.0   Married-civ-spouse   
48841   United-States   Bachelors           13.0             Divorced   

          Relationship  
0        Not-in-family  
1              Husband  
2        Not-in-family  
3              Husband  
4                 Wife  
...                ...  
48837        Own-child  
48838    Not-in-family  
48839   Other-relative  
48840          Husband  
48841        Own-child  

[48842 rows x 15 columns]

def altera_target(valor):
    if valor =='<=50K.':
        return '<=50K'
    if valor =='>50K.':
        return '>50K'
    return valor

dados_geral['Target'] = dados_geral['Target'].apply(altera_target)
dados_geral

# 3. Calcular a proporção entre pessoas com salários >50k em relação ao total com pessoas que difiram em sexo, raça e nível educacional. 
Por exemplo, ache qual é a porcentagem de mulheres com um salário maior que 50K e compare com a dos homens. O mesmo para pessoas de diferentes raças e níveis educacionais.

In [247]:
# Cálculo da porcentagem de homems e mulheres com salários >50k em relação ao total de pessoas.
df_Sex = dados_geral['Sex'].loc[dados_geral['Target'] == ' >50K'].value_counts()
df_Sex / 48841 * 100

 Male      6.664483
 Female    1.208001
Name: Sex, dtype: float64

In [248]:
# Cálculo da porcentagem de pessoas por raça com salários >50k em relação ao total de pessoas.
df_Race = dados_geral['Race'].loc[dados_geral['Target'] == ' >50K'].value_counts()
df_Race / 48841 * 100

 White                 7.143588
 Black                 0.366495
 Asian-Pac-Islander    0.272312
 Other                 0.051187
 Amer-Indian-Eskimo    0.038902
Name: Race, dtype: float64

In [249]:
# Cálculo da porcentagem de pessoas por níveis educacionais com salários >50k em relação ao total de pessoas.
df_Education = dados_geral['Education'].loc[dados_geral['Target'] == ' >50K'].value_counts()
df_Education / 48841 * 100

 Bachelors       2.233779
 HS-grad         1.695297
 Some-college    1.384083
 Masters         1.023730
 Prof-school     0.397207
 Assoc-voc       0.329641
 Assoc-acdm      0.303024
 Doctorate       0.255933
 11th            0.065519
 10th            0.051187
 7th-8th         0.045044
 12th            0.030712
 9th             0.028664
 5th-6th         0.022522
 1st-4th         0.004095
 Preschool       0.002047
Name: Education, dtype: float64